src/test/java/com/puppycrawl/tools/checkstyle/bdd/InlineConfigParser.java

```diff
@@ -344,21 +372,72 @@ else if (useFilteredViolations) {
+                                             int lineNo, String line)
+            throws ClassNotFoundException, CheckstyleException {
+        List<ModuleInputConfiguration> moduleLists = inputConfigBuilder.getChildrenModules();
+        final boolean specifyViolationMessage = moduleLists.size() == 1
+                && getNumberOfMessages(moduleLists.get(0).getModuleName()) > 1;
         final Matcher violationMatcher =
                 FILTERED_VIOLATION_PATTERN.matcher(line);
         final Matcher violationAboveMatcher =
                 FILTERED_VIOLATION_ABOVE_PATTERN.matcher(line);
         final Matcher violationBelowMatcher =
                 FILTERED_VIOLATION_BELOW_PATTERN.matcher(line);
         if (violationMatcher.matches()) {
-            inputConfigBuilder.addFilteredViolation(lineNo, violationMatcher.group(1));
+            final String violationMessage = violationMatcher.group(1);
+            checkWhetherViolationSpecified(specifyViolationMessage, violationMessage, lineNo);
+            inputConfigBuilder.addFilteredViolation(lineNo, violationMessage);
         }
         else if (violationAboveMatcher.matches()) {
-            inputConfigBuilder.addFilteredViolation(lineNo - 1, violationAboveMatcher.group(1));
+            final String violationMessage = violationAboveMatcher.group(1);
+            final int violationLineNum = lineNo - 1;
+            checkWhetherViolationSpecified(specifyViolationMessage, violationMessage,
+                    violationLineNum);
+            inputConfigBuilder.addFilteredViolation(violationLineNum, violationMessage);
         }
         else if (violationBelowMatcher.matches()) {
-            inputConfigBuilder.addFilteredViolation(lineNo + 1, violationBelowMatcher.group(1));
+            final String violationMessage = violationBelowMatcher.group(1);
+            final int violationLineNum = lineNo + 1;
+            checkWhetherViolationSpecified(specifyViolationMessage, violationMessage,
+                    violationLineNum);
+            inputConfigBuilder.addFilteredViolation(violationLineNum, violationMessage);
+        }
+    }
+
+    /**
+     * Gets the number of message keys in a check.
+     *
+     * @param className className
+     * @return number of message keys in a check
+     * @throws ClassNotFoundException classNotFoundException
+     */
+    private static long getNumberOfMessages(String className) throws ClassNotFoundException {
+        final Class<?> clazz = Class.forName(className);
+        final String messageInitials = "MSG_";
+        return Arrays.stream(clazz.getDeclaredFields())
+                .filter(field -> {
+                    final int modifiers = field.getModifiers();
+                    final String fieldName = field.getName();
+                    return fieldName.startsWith(messageInitials)
+                            && Modifier.isStatic(modifiers)
+                            && Modifier.isFinal(modifiers);
```

Comment:
Can be removed if there are no other fields in checks starting with `MSG_`

Reply:
All should be final. Good to keep.